ggml/src/ggml.c

```diff
@@ -5815,8 +5815,31 @@ static void ggml_visit_parents(struct ggml_cgraph * cgraph, struct ggml_tensor *
             (cgraph->order == GGML_CGRAPH_EVAL_ORDER_LEFT_TO_RIGHT) ? i :
             (cgraph->order == GGML_CGRAPH_EVAL_ORDER_RIGHT_TO_LEFT) ? (GGML_MAX_SRC-1-i) :
             /* unknown order, just fall back to using i*/ i;
-        if (node->src[k]) {
-            ggml_visit_parents(cgraph, node->src[k]);
+
+        struct ggml_tensor * s = node->src[k];
+        if (s) {
+            ggml_visit_parents(cgraph, s);
+
+            // Update the use count for this operand.
+            // Skip if it's a leaf node
+            if (!(s->op == GGML_OP_NONE && !(s->flags & GGML_TENSOR_FLAG_PARAM))) {
+                // the src can be the node itself (happens in ggml_cast)
+                if (s == node) {
+                    cgraph->use_counts[cgraph->n_nodes]++;
+                } else {
+                    // Search backward to find the src. This usually takes very few
+                    // (most often one) iteration(s). Probably comparable to hashing
+                    // on average..
+                    int j = cgraph->n_nodes - 1;
+                    for (; j >= 0; --j) {
+                        if (s == cgraph->nodes[j]) {
+                            break;
+                        }
+                    }
```

Comment:
Wouldn't it be better to use the hash table that is already in `ggml_cgraph`? It can be done by allocating an array of the same size as the hash table, and using `ggml_hash_find` to find the slot for the node.

Reply:
The search time can be significant, that's why the hash table was added, previously it scanned the previous nodes to determine if a node had already been visited. This implementation is not exactly what I meant, you can make `use_counts` have size `hash_size` and essentially use it as data attached to the hash table. The `memset` should also be avoided, it has a significant cost, which is also why the hash table has a bit field to determine which slots are used. I will take a look at this later and change it in this way.

Reply:
I see what you mean. I'll change it.

Reply:
done